### ISSUE TITLE
Suggest an index to the user when two dataflows are arranging the same thing

### ISSUE THREAD

**Issue by ggevay** [MaterializeInc/materialize#20802]
### Feature request

If we have two active dataflowa both arranging the same input, then we should suggest it to the user to index it.

This was happening recently to a customer: https://materializeinc.slack.com/archives/C02PPB50ZHS/p1690371972579099?thread_ts=1690294151.692909&cid=C02PPB50ZHS

cc @ggnall

**Comment by ggevay** [comment]
(Currently blocked by the need to record query plans in the catalog.)

**Comment by ggnall** [comment]
Blocked by: #20660

**Comment by ggevay** [comment]
Yes, https://github.com/MaterializeInc/materialize/issues/20660 would be one way to unblock this. However, I'm thinking that instead we might want to build on @aalexandrov's work that will add query plans to the catalog (https://github.com/MaterializeInc/materialize/issues/20652). We'll have to think this through carefully whether that is indeed the better approach. Some possible issues/questions with building this on the query log:
- We'd have to skip through a lot of one-off queries, which are irrelevant for this issue. Or will there be a fast way to ask the query log about only materialize views and indexes?
- The query log might be sampled to some degree, so maybe we's miss some relevant dataflows.
- Will the query log contain very old queries, or will it have some retention period? Some relevant dataflow might have been installed a long time ago.

**Comment by ggnall** [comment]
I see! 

Why is this? Not following.
> We'd have to skip through a lot of one-off queries, which are irrelevant for this issue

Not that I'm aware of. There was a recent request to add `collection_id`s, which may provide an index search? [Thread](https://materializeinc.slack.com/archives/C056M3001Q9/p1692046295199229?thread_ts=1691683618.682609&cid=C056M3001Q9) @umanwizard 
> Or will there be a fast way to ask the query log about only materialize views and indexes?

Confirming this will be sampled.
>The query log might be sampled to some degree, so maybe we'd miss some relevant dataflows.

Similarly, these will have a configurable retention period, e.g. 30 days.
>Will the query log contain very old queries, or will it have some retention period?

**Comment by ggevay** [comment]
>> We'd have to skip through a lot of one-off queries, which are irrelevant for this issue

> Why is this? Not following.

Sorry, maybe I didn't make it clear in the issue description (I've edited now) that I intended this issue for dealing with static dataflows, e.g., dataflows for indexes or mat views. If we look only at static dataflows, then this issue seems straightforward (after we'll have access to the plans from the catalog), because it's true in almost all cases that I can think of that if two static dataflows are arranging the same thing, then an index would improve things.

Doing similar suggestions for one-off selects has more open questions. First of all, it's less clear that two one-off selects arranging the same thing mean that the user definitely should create an index. This is because the index would use memory continuously, while the one-off selects might happen only sporadically. So it's not clear how many one-off selects (and over what time period) should arrange something before we give a hint that it should be indexed.

**Comment by ggnall** [comment]
Agree on focusing scope! One-off selects are still a problem for development workflows and ad hoc analysis, but I think this is the right prioritization given that static dataflows would be the ones powering operational workloads. 
>I intended this issue for dealing with static dataflows, e.g., dataflows for indexes or mat views.

**Comment by ggevay** [comment]
https://github.com/MaterializeInc/materialize/issues/20652 is resolved now, so this is good to go. I'll start working on it soon.

**Comment by aalexandrov** [comment]
[Related discussion in Slack](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1697790052129689).

**Comment by aalexandrov** [comment]
Putting some assorted thoughts here based on observations in customer environments.

TL;DR: a high-level policy might be:

- Identify views that are used more than once.
- Score higher views that are an MFP ending with a `Reduce` or `TopK`.
- Recommend index creation for such views that are referenced more than some number `k` (for example `k > 4`).
